From e029bbca0e1e4610684be040892a37f6abb132bb Mon Sep 17 00:00:00 2001
From: Guy Zadicario <guy.zadicario@intel.com>
Date: Wed, 1 Apr 2020 17:10:17 +0300
Subject: [PATCH v9 18/22] misc: nnpi: PCIe reset and error handling.

Handle PCIe reset and errors.
Inform all client applications about the error and stop any I/O
to the device until device is resumed.

Signed-off-by: Guy Zadicario <guy.zadicario@intel.com>
Reviewed-by: Vaibhav Agarwal <vaibhav.agarwal@intel.com>
---
 drivers/misc/intel-nnpi/device.c       | 113 +++++++++++++++++++-
 drivers/misc/intel-nnpi/device.h       |   5 +
 drivers/misc/intel-nnpi/device_sysfs.c |  35 ++++---
 drivers/misc/intel-nnpi/pcie.c         | 184 ++++++++++++++++++++++++++++++++-
 drivers/misc/intel-nnpi/pcie.h         |  17 +++
 5 files changed, 337 insertions(+), 17 deletions(-)

diff --git a/drivers/misc/intel-nnpi/device.c b/drivers/misc/intel-nnpi/device.c
index 0419ee4..2d278a6 100644
--- a/drivers/misc/intel-nnpi/device.c
+++ b/drivers/misc/intel-nnpi/device.c
@@ -528,6 +528,12 @@ static void handle_device_critical_error(struct nnp_device *nnpdev,
 				(u32)event_msg->obj_id;
 		nnpdev->counters.uncorr.os_crashed++;
 		break;
+	case NNP_IPC_ERROR_PCI_ERROR:
+		nnpdev->state |= NNP_DEVICE_PCI_ERROR;
+		nnpdev->pci_error = event_msg->event_val;
+		if (event_msg->event_val == NNP_PCIE_FATAL_ERROR)
+			nnpdev->counters.uncorr.os_crashed++;
+		break;
 	case NNP_IPC_ERROR_PROTOCOL_ERROR:
 		nnpdev->state |= NNP_DEVICE_PROTOCOL_ERROR;
 		break;
@@ -547,7 +553,7 @@ static void handle_device_critical_error(struct nnp_device *nnpdev,
 		nnpdev->state |= NNP_DEVICE_FATAL_ICE_ERROR;
 		break;
 	case NNP_IPC_ERROR_CARD_RESET:
-		if (nnpdev->state & NNP_DEVICE_HANG)
+		if ((nnpdev->state & NNP_DEVICE_HANG) && event_msg->event_val)
 			dev_info(nnpdev->dev,
 				 "hang detected - reset is needed\n");
 		break;
@@ -1407,6 +1413,20 @@ void nnpdrv_card_doorbell_value_changed(struct nnp_device *nnpdev,
 	queue_work(nnpdev->wq, &req->work);
 }
 
+static void notify_device_disconnect(struct nnp_device *nnpdev, bool is_reset)
+{
+	union c2h_event_report event;
+
+	/*
+	 * Report critical error event to all clients
+	 */
+	event.value = 0;
+	event.opcode = NNP_IPC_C2H_OP_EVENT_REPORT;
+	event.event_code = NNP_IPC_ERROR_CARD_RESET;
+	event.event_val = is_reset ? 1 : 0;
+	handle_event_report(nnpdev, &event, sizeof(event) / sizeof(u64));
+}
+
 /**
  * nnpdrv_device_destroy() - destroy nnp device object and wait until its removed
  * @nnpdev: The nnp device to be destroyed.
@@ -1436,6 +1456,11 @@ void nnpdrv_device_destroy(struct nnp_device *nnpdev, bool prepare_only)
 	dev_dbg(&nnpdev->nnp_pci->pdev->dev, "Destroying NNP-I device\n");
 
 	/*
+	 * Notify all client applications to stop using the device
+	 */
+	notify_device_disconnect(nnpdev, false);
+
+	/*
 	 * Decrement nnp_device refcount and wait until
 	 * all clients get disconnected, refcount reaches 0 and nnp_device
 	 * is released and freed
@@ -1530,3 +1555,89 @@ void nnpdrv_device_enable(struct nnp_device *nnpdev)
 	nnpdev->state |= NNP_DEVICE_CARD_ENABLED;
 	spin_unlock(&nnpdev->lock);
 }
+
+/**
+ * nnpdrv_device_pci_error_detected() - notification of pci error
+ * @nnpdev: nnp device
+ * @error_type: PCI error type, possible values in pcie.h
+ *
+ * This function is called by the PCI layer when a PCI error was detected.
+ */
+int nnpdrv_device_pci_error_detected(struct nnp_device *nnpdev,
+				     enum nnp_pcie_err error_type)
+{
+	union c2h_event_report event;
+
+	/*
+	 * report the event through event report path.
+	 */
+	event.value = 0;
+	event.opcode = NNP_IPC_C2H_OP_EVENT_REPORT;
+	event.event_code = NNP_IPC_ERROR_PCI_ERROR;
+	event.event_val = error_type;
+
+	handle_event_report(nnpdev, &event, sizeof(event) / sizeof(u64));
+
+	return 0;
+}
+
+/*
+ * nnpdrv_device_reset_prepare - notification that device is going to be reset
+ *
+ * @nnpdev: nnp device
+ * @is_hang: true if reset is due to device hang was detected.
+ *
+ * This function is called from PCI layer before a device reset flow starts.
+ */
+void nnpdrv_device_reset_prepare(struct nnp_device *nnpdev, bool is_hang)
+{
+	/* BAR2 can not longer be accessed */
+	spin_lock(&nnpdev->lock);
+	nnpdev->inbound_mem = NULL;
+	spin_unlock(&nnpdev->lock);
+
+	/*
+	 * Put the device in error state due to reset started.
+	 * Error condition will be cleared when boot state is changed.
+	 */
+	if (!is_hang) {
+		nnpdrv_set_boot_state(nnpdev, NNP_DEVICE_CARD_IN_RESET);
+
+		/* set host driver state to "Not ready" */
+		nnp_set_host_doorbell_value(nnpdev->nnp_pci, 0);
+	} else {
+		nnpdrv_set_boot_state(nnpdev, NNP_DEVICE_HANG);
+	}
+
+	/*
+	 * Remove and invalidate all message queues so that
+	 * no more messages will be sent to the h/w queue
+	 */
+	msg_sched_invalidate_all(nnpdev->cmdq_sched);
+
+	/*
+	 * Notify all client applications to stop using the device
+	 */
+	notify_device_disconnect(nnpdev, true);
+}
+
+/*
+ * nnpdrv_device_reset_done - device reset flow done notification
+ *
+ * @nnpdev: nnp device
+ *
+ * This function is called from PCI layer after device reset flow has
+ * successfully completed.
+ */
+void nnpdrv_device_reset_done(struct nnp_device *nnpdev)
+{
+	u64 membar_addr;
+	void *membar_vaddr;
+
+	/* re-enable the public command q */
+	msg_sched_queue_make_valid(nnpdev->public_cmdq);
+
+	/* recover BAR2 address after reset succeeded */
+	nnp_get_membar_addr(nnpdev->nnp_pci, &membar_addr, &membar_vaddr, NULL);
+	nnpdev->inbound_mem = (union nnp_inbound_mem *)membar_vaddr;
+}
diff --git a/drivers/misc/intel-nnpi/device.h b/drivers/misc/intel-nnpi/device.h
index 671317e..9075a79 100644
--- a/drivers/misc/intel-nnpi/device.h
+++ b/drivers/misc/intel-nnpi/device.h
@@ -183,6 +183,8 @@ struct nnp_device_counters {
  * @num_active_contexts: number of active inference contexts running.
  * @curr_boot_state: last boot state field received from device doorbell reg
  * @card_doorbell_val: last received device doorbell register value.
+ * @pci_error: non-zero if pci error has occurred, possible values are one of
+ *             NNP_PCIE_* values defined in pcie.h
  * @host_crash_dump: allocated dma page info to host device crash dump
  * @query_version_work: work struct used to schedule processing of version
  *                      reply response message arrived from device.
@@ -250,6 +252,7 @@ struct nnp_device {
 	unsigned short chan_protocol_version;
 	unsigned int   num_active_contexts;
 	u32            card_doorbell_val;
+	u32            pci_error;
 
 	struct host_crash_dump    host_crash_dump;
 	struct query_version_work query_version_work;
@@ -289,6 +292,8 @@ static inline int nnpdrv_msg_sched_queue_add_msg(struct msg_sched_queue *queue,
 void nnpdrv_device_put(struct nnp_device *nnpdev);
 
 void nnpdrv_set_boot_state(struct nnp_device *nnpdev, u32 mask);
+void nnpdrv_device_disable(struct nnp_device *nnpdev);
+void nnpdrv_device_enable(struct nnp_device *nnpdev);
 
 struct nnpdrv_cmd_chan *nnpdrv_find_channel(struct nnp_device *nnpdev,
 					    u16 chan_id);
diff --git a/drivers/misc/intel-nnpi/device_sysfs.c b/drivers/misc/intel-nnpi/device_sysfs.c
index beb65ab..2027cf6 100644
--- a/drivers/misc/intel-nnpi/device_sysfs.c
+++ b/drivers/misc/intel-nnpi/device_sysfs.c
@@ -145,34 +145,41 @@ void nnpdrv_device_sysfs_get_state_strings(struct nnp_device *nnpdev,
 		goto find_state;
 
 	/* Find failure Reason*/
-	if (nnpdev->state & NNP_DEVICE_FAILED_VERSION)
+	if (nnpdev->state &
+	    (NNP_DEVICE_FAILED_VERSION | NNP_DEVICE_PROTOCOL_ERROR)) {
 		*fail_reason = "version Mismatch";
-	else if (nnpdev->state & NNP_DEVICE_CAPSULE_EXPECTED)
+	} else if (nnpdev->state & NNP_DEVICE_CAPSULE_EXPECTED) {
 		*fail_reason = "Bios Update Required";
-	else if (nnpdev->state & NNP_DEVICE_CAPSULE_FAILED)
+	} else if (nnpdev->state & NNP_DEVICE_CAPSULE_FAILED) {
 		*fail_reason = "Bios Update Failed";
-	else if (nnpdev->state & NNP_DEVICE_CURRUPTED_BOOT_IMAGE)
+	} else if (nnpdev->state & NNP_DEVICE_CURRUPTED_BOOT_IMAGE) {
 		*fail_reason = "Boot/Capsule image verification Failed";
-	else if (nnpdev->state & NNP_DEVICE_BOOT_FAILED)
+	} else if (nnpdev->state & NNP_DEVICE_BOOT_FAILED) {
 		*fail_reason = "Boot failed";
-	else if (nnpdev->state & NNP_DEVICE_HOST_DRIVER_ERROR)
+	} else if (nnpdev->state & NNP_DEVICE_HOST_DRIVER_ERROR) {
 		*fail_reason = "Driver Error";
-	else if (nnpdev->state & NNP_DEVICE_KERNEL_CRASH)
+	} else if (nnpdev->state & NNP_DEVICE_KERNEL_CRASH) {
 		*fail_reason = "OS Crash";
-	else if (nnpdev->state & NNP_DEVICE_CARD_IN_RESET)
+	} else if (nnpdev->state & NNP_DEVICE_PCI_ERROR) {
+		if (nnpdev->pci_error == NNP_PCIE_LINK_RETRAIN_REQUIRED)
+			*fail_reason = "PCI Error (rescan required)";
+		else
+			*fail_reason = "PCI Error";
+	} else if (nnpdev->state & NNP_DEVICE_CARD_IN_RESET) {
 		*fail_reason = "Reset in progress";
-	else if (nnpdev->state & NNP_DEVICE_FATAL_MCE_ERROR)
+	} else if (nnpdev->state & NNP_DEVICE_FATAL_MCE_ERROR) {
 		*fail_reason = "Fatal MCE Error";
-	else if (nnpdev->state & NNP_DEVICE_FATAL_DRAM_ECC_ERROR)
+	} else if (nnpdev->state & NNP_DEVICE_FATAL_DRAM_ECC_ERROR) {
 		*fail_reason = "Fatal DRAM ECC Error";
-	else if (nnpdev->state & NNP_DEVICE_FATAL_ICE_ERROR)
+	} else if (nnpdev->state & NNP_DEVICE_FATAL_ICE_ERROR) {
 		*fail_reason = "Fatal ICE execution error";
-	else if (nnpdev->state & NNP_DEVICE_FATAL_DMA_ERROR)
+	} else if (nnpdev->state & NNP_DEVICE_FATAL_DMA_ERROR) {
 		*fail_reason = "Fatal DMA hang error";
-	else if (nnpdev->state & NNP_DEVICE_HANG)
+	} else if (nnpdev->state & NNP_DEVICE_HANG) {
 		*fail_reason = "Device Not Responding (Hang)";
-	else
+	} else {
 		*fail_reason = "None";
+	}
 
 find_state:
 	if (!state)
diff --git a/drivers/misc/intel-nnpi/pcie.c b/drivers/misc/intel-nnpi/pcie.c
index b4af404..e1f7057 100644
--- a/drivers/misc/intel-nnpi/pcie.c
+++ b/drivers/misc/intel-nnpi/pcie.c
@@ -6,6 +6,7 @@
 
 #define pr_fmt(fmt)   KBUILD_MODNAME ": " fmt
 
+#include <linux/aer.h>
 #include <linux/bitfield.h>
 #include <linux/delay.h>
 #include <linux/dev_printk.h>
@@ -256,6 +257,8 @@ static void nnp_free_interrupts(struct nnp_pci_device *nnp_pci,
  *            advance its command queue read pointer. The caller may wait
  *            for this counter to be advanced past this point before calling
  *            this function again to re-try the write.
+ * * -ENODEV: device pci is not initialized. Can happen when device reset is
+ *            in progress.
  */
 static int nnp_cmdq_write_mesg_nowait(struct nnp_pci_device *nnp_pci,
 				      u64 *msg, u32 size,
@@ -266,6 +269,9 @@ static int nnp_cmdq_write_mesg_nowait(struct nnp_pci_device *nnp_pci,
 	unsigned long flags;
 	int i;
 
+	if (!nnp_pci->initted)
+		return -ENODEV;
+
 	if (!size)
 		return 0;
 
@@ -344,6 +350,9 @@ static bool check_read_count(struct nnp_pci_device *nnp_pci, u32 count)
  *
  * Return:
  * * 0: Success, command has been written
+ * * -ENODEV: device pci is not initialized. Can happen when device reset is
+ *            in progress, or started while waiting for space in the command
+ *            queue.
  */
 int nnp_cmdq_write_mesg(struct nnp_pci_device *nnp_pci,
 			u64 *msg, u32 size, u64 *timed_wait)
@@ -362,7 +371,8 @@ int nnp_cmdq_write_mesg(struct nnp_pci_device *nnp_pci,
 
 	while (rc == -EAGAIN) {
 		rc = wait_event_interruptible(nnp_pci->card_status_wait,
-					      check_read_count(nnp_pci, rcnt));
+					      check_read_count(nnp_pci, rcnt) ||
+					      !nnp_pci->initted);
 		if (rc)
 			break;
 
@@ -381,6 +391,10 @@ int nnp_cmdq_write_mesg(struct nnp_pci_device *nnp_pci,
 
 int nnp_cmdq_flush(struct nnp_pci_device *nnp_pci)
 {
+	/* Can happen during device reset */
+	if (!nnp_pci->initted)
+		return -ENODEV;
+
 	nnp_mmio_write(nnp_pci, ELBI_COMMAND_PCI_CONTROL,
 		       ELBI_COMMAND_PCI_CONTROL_FLUSH_MASK);
 
@@ -389,6 +403,10 @@ int nnp_cmdq_flush(struct nnp_pci_device *nnp_pci)
 
 int nnp_set_host_doorbell_value(struct nnp_pci_device *nnp_pci, u32 value)
 {
+	/* Can happen during device reset */
+	if (!nnp_pci->initted)
+		return -ENODEV;
+
 	/*
 	 * The SELF_RESET bit is set only by the h/w layer,
 	 * do not allow higher layer to set it
@@ -426,6 +444,10 @@ int nnp_fw_update_complete(struct nnp_pci_device *nnp_pci, bool start)
 	u32 host_db_val;
 	bool need;
 
+	/* in case we are called during device reset/remove */
+	if (!nnp_pci->initted)
+		return -ENODEV;
+
 	card_state = nnp_mmio_read(nnp_pci, ELBI_CPU_STATUS_2);
 	need = (FIELD_GET(ELBI_CPU_STATUS_2_FLR_MODE_MASK, card_state) == 1);
 
@@ -542,6 +564,11 @@ static int nnp_init_pci_device(struct nnp_pci_device *nnp_pci)
 		goto unmap_mem_bar;
 	}
 
+	rc = pci_enable_pcie_error_reporting(pdev);
+	if (rc)
+		dev_dbg(&nnp_pci->pdev->dev,
+			"pci_enable_pcie_error_reporting returned %d\n", rc);
+
 	/*
 	 * done setting up the new pci device,
 	 * add it to the set of NNP-I devices
@@ -583,6 +610,7 @@ static int nnp_init_pci_device(struct nnp_pci_device *nnp_pci)
 		       ELBI_PCI_MSI_MASK,
 		       ~(card_status_int_mask | card_status_threaded_mask));
 
+	nnp_pci->initted = true;
 	return 0;
 
 free_interrupts:
@@ -606,6 +634,10 @@ static int nnp_init_pci_device(struct nnp_pci_device *nnp_pci)
 static void nnp_fini_pci_device(struct nnp_pci_device *nnp_pci,
 				bool release_regions)
 {
+	if (!nnp_pci->initted)
+		return;
+
+	nnp_pci->initted = false;
 	wake_up_all(&nnp_pci->card_status_wait);
 	nnp_free_interrupts(nnp_pci, nnp_pci->pdev);
 	if (release_regions && nnp_pci->mem_enabled) {
@@ -642,6 +674,7 @@ static int nnp_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 	init_waitqueue_head(&nnp_pci->card_status_wait);
 	spin_lock_init(&nnp_pci->cmdq_lock);
 	spin_lock_init(&nnp_pci->irq_lock);
+	mutex_init(&nnp_pci->remove_reset_mutex);
 
 	rc = nnp_init_pci_device(nnp_pci);
 	if (rc)
@@ -670,6 +703,8 @@ static void nnp_remove(struct pci_dev *pdev)
 	if (!nnp_pci)
 		return;
 
+	mutex_lock(&nnp_pci->remove_reset_mutex);
+
 	nnp_pci->removing = true;
 
 	/* inform card that host driver is down */
@@ -678,12 +713,156 @@ static void nnp_remove(struct pci_dev *pdev)
 	nnpdrv_device_destroy(nnp_pci->nnpdev, true);
 	nnp_fini_pci_device(nnp_pci, true);
 	nnpdrv_device_destroy(nnp_pci->nnpdev, false);
+	mutex_unlock(&nnp_pci->remove_reset_mutex);
 
 	pci_set_drvdata(nnp_pci->pdev, NULL);
 	pci_dev_put(nnp_pci->pdev);
 	kfree(nnp_pci);
 }
 
+static pci_ers_result_t nnp_pci_err_error_detected(struct pci_dev *dev,
+						   pci_channel_state_t error)
+{
+	struct nnp_pci_device *nnp_pci = NULL;
+
+	dev_err(&nnp_pci->pdev->dev, "PCI error detected error=%d\n", error);
+
+	nnp_pci = pci_get_drvdata(dev);
+	if (!nnp_pci)
+		return PCI_ERS_RESULT_NONE;
+
+	if (error == pci_channel_io_normal) {
+		/* non-fatal error */
+
+		/* report the event upstream */
+		nnpdrv_device_pci_error_detected(nnp_pci->nnpdev,
+						 NNP_PCIE_NON_FATAL_ERROR);
+
+		/* no need to reset the PCI bus */
+		return PCI_ERS_RESULT_CAN_RECOVER;
+	} else if (error == pci_channel_io_frozen) {
+		/* fatal error */
+
+		/* report the event upstream */
+		nnpdrv_device_pci_error_detected(nnp_pci->nnpdev,
+						 NNP_PCIE_FATAL_ERROR);
+
+		/* need to reset the link */
+		return PCI_ERS_RESULT_NEED_RESET;
+	} else if (error == pci_channel_io_perm_failure) {
+		/* cannot be recovered */
+
+		/* report the event upstream */
+		nnpdrv_device_pci_error_detected(nnp_pci->nnpdev,
+						 NNP_PCIE_PERMANENT_FAILURE);
+
+		return PCI_ERS_RESULT_DISCONNECT;
+	}
+
+	return PCI_ERS_RESULT_NONE;
+}
+
+static pci_ers_result_t nnp_pci_err_mmio_enabled(struct pci_dev *dev)
+{
+	dev_err(&dev->dev, "PCI error mmio_enabled\n");
+	return PCI_ERS_RESULT_NONE;
+}
+
+static pci_ers_result_t nnp_pci_err_slot_reset(struct pci_dev *dev)
+{
+	u32 cmd;
+
+	dev_err(&dev->dev, "PCI error slot_reset\n");
+
+	pci_read_config_dword(dev, 0x4, &cmd);
+	if (cmd != 0xffffffff)
+		return PCI_ERS_RESULT_RECOVERED;
+	else
+		return PCI_ERS_RESULT_DISCONNECT;
+}
+
+static void nnp_pci_err_resume(struct pci_dev *dev)
+{
+	struct nnp_pci_device *nnp_pci = NULL;
+
+	nnp_pci = pci_get_drvdata(dev);
+	if (!nnp_pci)
+		return;
+
+	dev_err(&nnp_pci->pdev->dev, "PCI error resume\n");
+}
+
+static void nnp_reset_prepare(struct pci_dev *dev)
+{
+	struct nnp_pci_device *nnp_pci = NULL;
+
+	nnp_pci = pci_get_drvdata(dev);
+
+	dev_dbg(&nnp_pci->pdev->dev, "reset_prepare\n");
+	mutex_lock(&nnp_pci->remove_reset_mutex);
+	if (nnp_pci->removing)
+		return; /* unlock will happen on nnp_reset_done */
+
+	nnpdrv_device_reset_prepare(nnp_pci->nnpdev, false);
+
+	nnp_fini_pci_device(nnp_pci, false);
+}
+
+static void nnp_reset_done(struct pci_dev *dev)
+{
+	struct nnp_pci_device *nnp_pci = NULL;
+	int rc = 0;
+	u32 cmd;
+	int t = 30;
+
+	nnp_pci = pci_get_drvdata(dev);
+
+	dev_dbg(&nnp_pci->pdev->dev, "reset_done\n");
+
+	if (nnp_pci->removing)
+		goto done; // device removed before reset has started
+
+	do {
+		pci_read_config_dword(dev, 0x4, &cmd);
+		dev_dbg(&nnp_pci->pdev->dev,
+			"config after reset t=%d cmd0 = 0x%x\n", t, cmd);
+		if (cmd != 0xffffffff)
+			break;
+		msleep(100);
+	} while (t-- > 0);
+
+	if (cmd != 0xffffffff)
+		rc = nnp_init_pci_device(nnp_pci);
+
+	if (cmd == 0xffffffff || rc) {
+		u32 err = (cmd == 0xffffffff || rc == -EIO ?
+			   NNP_PCIE_LINK_RETRAIN_REQUIRED :
+			   NNP_PCIE_PERMANENT_FAILURE);
+
+		dev_err(&nnp_pci->pdev->dev,
+			"failed to initialize pci device after FLR/Reset!!\n");
+		if (err == NNP_PCIE_LINK_RETRAIN_REQUIRED)
+			dev_err(&nnp_pci->pdev->dev,
+				"Remove and Rescan device may help.\n");
+		if (nnp_pci->nnpdev)
+			nnpdrv_device_pci_error_detected(nnp_pci->nnpdev, err);
+	} else {
+		nnpdrv_device_reset_done(nnp_pci->nnpdev);
+	}
+
+done:
+	mutex_unlock(&nnp_pci->remove_reset_mutex);
+}
+
+static struct pci_error_handlers nnp_pci_err_handlers = {
+	.error_detected = nnp_pci_err_error_detected,
+	.mmio_enabled = nnp_pci_err_mmio_enabled,
+	.slot_reset = nnp_pci_err_slot_reset,
+	.reset_prepare = nnp_reset_prepare,
+	.reset_done = nnp_reset_done,
+	.resume = nnp_pci_err_resume,
+};
+
 static const struct pci_device_id nnp_pci_tbl[] = {
 	{PCI_DEVICE(NNP_PCI_VENDOR_ID, NNP_PCI_DEVICE_ID)},
 	/* required last entry */
@@ -694,7 +873,8 @@ static void nnp_remove(struct pci_dev *pdev)
 	.name = nnp_driver_name,
 	.id_table = nnp_pci_tbl,
 	.probe = nnp_probe,
-	.remove = nnp_remove
+	.remove = nnp_remove,
+	.err_handler = &nnp_pci_err_handlers
 };
 
 int nnpdrv_pci_module_init(void)
diff --git a/drivers/misc/intel-nnpi/pcie.h b/drivers/misc/intel-nnpi/pcie.h
index d51a473..747fc48 100644
--- a/drivers/misc/intel-nnpi/pcie.h
+++ b/drivers/misc/intel-nnpi/pcie.h
@@ -41,6 +41,7 @@ struct nnp_memdesc {
  * @nnpdev: pointer to the "logical" device structure of this NNP-I device.
  *          This is to separate NNP-I common device management from hw device
  *          management (which potentially be not a pci device).
+ * @remove_reset_mutex: serialize device remove and reset flows.
  * @mmio: device's BAR0 mapping info
  * @mem_bar: device's BAR2 mapping info, this is the "inbound memory region"
  *           This device memory region is described in
@@ -68,11 +69,14 @@ struct nnp_memdesc {
  *               handler.
  * @cmd_read_update_count: number of times the device has updated its read
  *                         pointer to the device command queue.
+ * @initted: true if PCI is enabled and interrupts are attached and
+ *           enabled. This will be false during device reset flow.
  * @removing: true if device remove is in progress.
  */
 struct nnp_pci_device {
 	struct pci_dev    *pdev;
 	struct nnp_device *nnpdev;
+	struct mutex    remove_reset_mutex; // serialize reset and remove flows
 
 	struct nnp_memdesc        mmio;
 	struct nnp_memdesc        mem_bar; /* the inbound memory region */
@@ -91,6 +95,8 @@ struct nnp_pci_device {
 	u32             card_status;
 	u32             cmd_read_update_count;
 
+	u32             initted;
+
 	bool            removing;
 };
 
@@ -109,6 +115,13 @@ int nnp_get_membar_addr(struct nnp_pci_device *nnp_pci,
 			phys_addr_t *out_phy_addr, void __iomem **out_vaddr,
 			size_t *out_len);
 
+enum nnp_pcie_err {
+	NNP_PCIE_NON_FATAL_ERROR = 1,
+	NNP_PCIE_FATAL_ERROR,
+	NNP_PCIE_PERMANENT_FAILURE,
+	NNP_PCIE_LINK_RETRAIN_REQUIRED
+};
+
 /*
  * Functions implemented by the nnp "device" layer,
  * called by the nnp "pci" layer
@@ -119,6 +132,10 @@ void nnpdrv_card_doorbell_value_changed(struct nnp_device *nnpdev,
 					u32 doorbell_val);
 void nnpdrv_device_process_messages(struct nnp_device *nnpdev, u64 *hw_msg,
 				    unsigned int hw_nof_msg);
+int nnpdrv_device_pci_error_detected(struct nnp_device *nnpdev,
+				     enum nnp_pcie_err error_type);
+void nnpdrv_device_reset_prepare(struct nnp_device *nnpdev, bool is_hang);
+void nnpdrv_device_reset_done(struct nnp_device *nnpdev);
 
 int nnpdrv_pci_module_init(void);
 void nnpdrv_pci_module_fini(void);
-- 
1.8.3.1

